File: packages/recovery-capacitor/ios/Sources/KeywardRecoveryPlugin/KeywardRecoveryPlugin.swift
import AuthenticationServices
import Capacitor
import Foundation
import UIKit
#if canImport(KeywardRecovery)
import KeywardRecovery
#endif

@objc(KeywardRecoveryPlugin)
public class KeywardRecoveryPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "KeywardRecoveryPlugin"
    public let jsName = "KeywardRecovery"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "capabilities", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prfRegister", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "prfAssert", returnType: CAPPluginReturnPromise),
    ]

    @objc func capabilities(_ call: CAPPluginCall) {
        let caps = KeywardRecovery.capabilities()
        call.resolve([
            "platformAuthenticator": caps.platformAuthenticator,
            "prfSupported": caps.prfSupported,
            "prfAtCreate": caps.prfAtCreate,
        ])
    }

    @objc func prfRegister(_ call: CAPPluginCall) {
        guard
            let rpId = call.getString("rpId"),
            let userId = call.getString("userId"),
            let userName = call.getString("userName"),
            let challenge = call.getString("challenge"),
            let saltFirst = call.getString("saltFirst")
        else {
            call.reject("Missing required parameters")
            return
        }
        guard #available(iOS 18.0, *) else {
            call.reject("PRF requires iOS 18+")
            return
        }
        configureAnchor()
        Task {
            do {
                let result = try await KeywardRecovery.register(
                    rpId: rpId,
                    userId: userId,
                    userName: userName,
                    challenge: challenge,
                    saltFirst: saltFirst
                )
                call.resolve([
                    "credentialId": result.credentialId,
                    "prfFirst": result.prfFirst as Any,
                ])
            } catch {
                call.reject("prfRegister failed", nil, error)
            }
        }
    }

    @objc func prfAssert(_ call: CAPPluginCall) {
        guard
            let rpId = call.getString("rpId"),
            let challenge = call.getString("challenge"),
            let saltFirst = call.getString("saltFirst")
        else {
            call.reject("Missing required parameters")
            return
        }
        let credentialId = call.getString("credentialId")
        guard #available(iOS 18.0, *) else {
            call.reject("PRF requires iOS 18+")
            return
        }
        configureAnchor()
        Task {
            do {
                let result = try await KeywardRecovery.assert(
                    rpId: rpId,
                    challenge: challenge,
                    saltFirst: saltFirst,
                    credentialId: credentialId
                )
                call.resolve([
                    "credentialId": result.credentialId,
                    "prfFirst": result.prfFirst,
                ])
            } catch {
                call.reject("prfAssert failed", nil, error)
            }
        }
    }

    /// Give the native facade a presentation anchor (the app's key window).
    private func configureAnchor() {
        KeywardRecovery.anchorProvider = { [weak self] in
            self?.bridge?.viewController?.view.window ?? UIWindow()
        }
    }
}

File: packages/recovery-capacitor/android/src/main/java/org/keyward/recovery/capacitor/KeywardRecoveryPlugin.java
package org.keyward.recovery.capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.keyward.recovery.KeywardRecoveryPrf;
import org.keyward.recovery.PrfAssertResult;
import org.keyward.recovery.PrfCapabilities;
import org.keyward.recovery.PrfRegisterResult;

@CapacitorPlugin(name = "KeywardRecovery")
public class KeywardRecoveryPlugin extends Plugin {

    private KeywardRecoveryPrf prf;

    @Override
    public void load() {
        prf = new CredentialManagerPrfBackend(getActivity());
    }

    @PluginMethod
    public void capabilities(PluginCall call) {
        PrfCapabilities caps = prf.capabilities();
        JSObject result = new JSObject();
        result.put("platformAuthenticator", caps.platformAuthenticator);
        result.put("prfSupported", caps.prfSupported);
        result.put("prfAtCreate", caps.prfAtCreate);
        call.resolve(result);
    }

    @PluginMethod
    public void prfRegister(PluginCall call) {
        String rpId = call.getString("rpId");
        String userId = call.getString("userId");
        String userName = call.getString("userName");
        String challenge = call.getString("challenge");
        String saltFirst = call.getString("saltFirst");
        if (rpId == null || userId == null || userName == null || challenge == null || saltFirst == null) {
            call.reject("Missing required parameters");
            return;
        }
        prf.register(rpId, userId, userName, challenge, saltFirst,
                new KeywardRecoveryPrf.Callback<PrfRegisterResult>() {
                    @Override
                    public void onSuccess(PrfRegisterResult result) {
                        JSObject r = new JSObject();
                        r.put("credentialId", result.credentialId);
                        r.put("prfFirst", result.prfFirst);
                        call.resolve(r);
                    }

                    @Override
                    public void onError(Throwable error) {
                        call.reject(
                                "prfRegister failed",
                                error instanceof Exception ? (Exception) error : new Exception(error));
                    }
                });
    }

    @PluginMethod
    public void prfAssert(PluginCall call) {
        String rpId = call.getString("rpId");
        String challenge = call.getString("challenge");
        String saltFirst = call.getString("saltFirst");
        String credentialId = call.getString("credentialId");
        if (rpId == null || challenge == null || saltFirst == null) {
            call.reject("Missing required parameters");
            return;
        }
        prf.assertPrf(rpId, challenge, saltFirst, credentialId,
                new KeywardRecoveryPrf.Callback<PrfAssertResult>() {
                    @Override
                    public void onSuccess(PrfAssertResult result) {
                        JSObject r = new JSObject();
                        r.put("credentialId", result.credentialId);
                        r.put("prfFirst", result.prfFirst);
                        call.resolve(r);
                    }

                    @Override
                    public void onError(Throwable error) {
                        call.reject(
                                "prfAssert failed",
                                error instanceof Exception ? (Exception) error : new Exception(error));
                    }
                });
    }
}
